File: windows/src/GoalMaker.Infrastructure/Activity/PostgrestActivityLog.cs
using System.Globalization;
using System.Text;
using System.Text.Json.Nodes;
using GoalMaker.Core.Activity;
using GoalMaker.Core.Sync;
using GoalMaker.Infrastructure.Postgrest;

namespace GoalMaker.Infrastructure.Activity;

/// <summary><see cref="IActivityLog"/> over PostgREST: the <c>activity_log</c> table and the <c>undo_activity</c> function (0007).</summary>
public sealed class PostgrestActivityLog(PostgrestHttp postgrest) : IActivityLog
{
    public async Task<IReadOnlyList<ActivityEntry>> RecentAsync(int limit = 60, CancellationToken cancellationToken = default)
    {
        using var request = postgrest.Request(
            HttpMethod.Get,
            $"activity_log?select=id,entity,entity_id,action,actor,before,after,created_at,undone_at&order=id.desc&limit={limit}");
        var text = await postgrest.SendAsync(request, cancellationToken).ConfigureAwait(false);
        return [.. (JsonNode.Parse(text) as JsonArray ?? []).OfType<JsonObject>().Select(row => new ActivityEntry(
            (long)row["id"]!,
            (string)row["entity"]!,
            (string)row["entity_id"]!,
            (string)row["action"]!,
            (string)row["actor"]!,
            row["before"] as JsonObject,
            row["after"] as JsonObject ?? [],
            Instant((string?)row["created_at"]) ?? DateTimeOffset.UnixEpoch,
            Instant((string?)row["undone_at"])))];
    }

    public async Task<UndoOutcome> UndoAsync(long entryId, CancellationToken cancellationToken = default)
    {
        using var request = postgrest.Request(HttpMethod.Post, "rpc/undo_activity");
        request.Content = new StringContent(new JsonObject { ["entry_id"] = entryId }.ToJsonString(), Encoding.UTF8, "application/json");
        try
        {
            await postgrest.SendAsync(request, cancellationToken).ConfigureAwait(false);
            return UndoOutcome.Undone;
        }
        catch (RemoteRejectedException refused)
        {
            return refused.Code switch
            {
                "40001" => UndoOutcome.ChangedSince,
                "55000" => UndoOutcome.AlreadyUndone,
                _ => UndoOutcome.NotPossible,
            };
        }
    }

    private static DateTimeOffset? Instant(string? text) =>
        text is not null && DateTimeOffset.TryParse(text, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var instant)
            ? instant
            : null;
}

File: android/app/src/main/kotlin/com/goalmaker/app/data/activity/PostgrestActivityLog.kt
package com.goalmaker.app.data.activity

import com.goalmaker.app.application.activity.ActivityEntry
import com.goalmaker.app.application.activity.ActivityLog
import com.goalmaker.app.application.activity.UndoOutcome
import com.goalmaker.app.application.sync.RemoteRejectedException
import com.goalmaker.app.data.supabase.PostgrestHttp
import com.goalmaker.app.domain.sync.SyncRules
import io.ktor.client.request.parameter
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import java.time.Instant
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/** [ActivityLog] over PostgREST: the `activity_log` table and the `undo_activity` function (0007). */
class PostgrestActivityLog(private val postgrest: PostgrestHttp) : ActivityLog {
    override suspend fun recent(limit: Int): List<ActivityEntry> {
        val text = postgrest.send(HttpMethod.Get, "activity_log") {
            parameter("select", "id,entity,entity_id,action,actor,before,after,created_at,undone_at")
            parameter("order", "id.desc")
            parameter("limit", limit)
        }
        // A row missing what an entry needs is passed over: a stranger's row never empties the screen.
        return (Json.parseToJsonElement(text) as? JsonArray).orEmpty().filterIsInstance<JsonObject>().mapNotNull { row ->
            ActivityEntry(
                id = (row["id"] as? JsonPrimitive)?.longOrNull ?: return@mapNotNull null,
                entity = row.text("entity") ?: return@mapNotNull null,
                entityId = row.text("entity_id") ?: return@mapNotNull null,
                action = row.text("action") ?: return@mapNotNull null,
                actor = row.text("actor").orEmpty(),
                before = row["before"] as? JsonObject,
                after = row["after"] as? JsonObject ?: JsonObject(emptyMap()),
                createdAt = row.text("created_at")?.let(SyncRules::instantOf) ?: Instant.EPOCH,
                undoneAt = row.text("undone_at")?.let(SyncRules::instantOf),
            )
        }
    }

    override suspend fun undo(entryId: Long): UndoOutcome = try {
        postgrest.send(HttpMethod.Post, "rpc/undo_activity") {
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("entry_id", entryId) }.toString())
        }
        UndoOutcome.UNDONE
    } catch (refused: RemoteRejectedException) {
        when (refused.code) {
            "40001" -> UndoOutcome.CHANGED_SINCE
            "55000" -> UndoOutcome.ALREADY_UNDONE
            else -> UndoOutcome.NOT_POSSIBLE
        }
    }

    private fun JsonObject.text(name: String): String? = (this[name] as? JsonPrimitive)?.takeUnless { it == JsonNull }?.content
}
